### ISSUE TITLE
Minor Cleanup/Refactoring Issues

### ISSUE THREAD

**Issue by mrks** [hyrise/hyrise#1912]
Here, we want to collect minor things that caught our eyes that to not warrant their own issue but should be refactored at some point

**Comment by mrks** [comment]
Inconsistent use of shared_ptr/unique_ptr/immediate member in segments: Why does DictionarySegment store a shared_ptrs to the dictionary?

**Comment by mrks** [comment]
FrameOfReferenceSegment: Replace `std::enable_if_t<encoding_supports_data_type` with `static_assert`

**Comment by tobodner** [comment]
Now that we upgraded cpplint, we should reassess our NOLINT annotations.

**Comment by j-tr** [comment]
when trying to replace `enable_if` with `static_assert` i came across lots of methods who actually try to instanciate a FrameOfReferenceSegment for unsupported types and apparently rely on the hiding effect of `enable_if`. so it might not be possible to use static_assert here, since the enable_if is not only a security check.
I read about concepts in c++20 and thought, this could be a nice way to handle it, but since we are working with g++-9 so far, this is not yet implemented. 
any other thoughts on how this could be done without the need of changing all methods that wrongly instanciate a FrameOfReferenceSegment?

**Comment by mrks** [comment]
Could you give an example?

**Comment by j-tr** [comment]
for example within the lambdas in `resolve_data_and_segment_type` in `JoinNestedLoop::_join_two_untyped_segments` and `ColumnVsColumnTableScanImpl::scan_chunk` or during `EXPLICITLY_INSTANTIATE_DATA_TYPES(CreateAnySegmentIterable)` in `any_segment_iterable.cpp`. I don't fully understand why there are wrong instatiations but i'll investigate further

**Comment by mrks** [comment]
This is a mean one. Here's my current interpretation (about which I am not really sure). Let's break the error message down:

```
../src/lib/storage/frame_of_reference_segment.hpp:32:1: error: static_assert failed due to requirement 'encoding_supports_data_type(enum_c, hana::type_c<long>)' "bla"
static_assert(encoding_supports_data_type(enum_c<EncodingType, EncodingType::FrameOfReference>, hana::type_c<T>), "bla");
^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../src/lib/storage/segment_iterables/any_segment_iterable.cpp:16:75: note: in instantiation of template class 'opossum::FrameOfReferenceSegment<long>' requested here
      const auto actual_iterable = create_iterable_from_segment<T, false>(segment);
                                                                          ^
../src/lib/storage/segment_iterables/any_segment_iterable.cpp:16:36: note: while substituting deduced template arguments into function template 'create_iterable_from_segment' [with T = long, EraseSegmentType = false]
      const auto actual_iterable = create_iterable_from_segment<T, false>(segment);
                                   ^
../src/lib/resolve_type.hpp:205:5: note: in instantiation of function template specialization 'opossum::detail::CreateAnySegmentIterable<long>::create(const opossum::BaseSegment &)::(anonymous class)::operator()<opossum::ValueSegment<long> >' requested here
    func(*value_segment);
    ^
```

* Starting from the bottom, we see `CreateAnySegmentIterable<T>::create` and, within that, an anonymous class (the lambda used in `resolve_segment_type`).
* The lambda is called with `ValueSegment<long>` as the type of `segment`.
* Internally, it calls `create_iterable_from_segment`, an [overloaded named function](https://en.cppreference.com/w/cpp/language/overload_resolution).
* Theoretically, there would be no need for `create_iterable_from_segment(FrameOfReferenceSegment<long>)` to exist at this point.
* "For each function template, if the argument deduction and checking succeeds, the template-arguments (deduced and/or explicit) are used to synthesize the declaration of a single function template specialization which is added to the candidate functions set to be used in overload resolution." [[temp.over]](https://timsong-cpp.github.io/cppwp/temp.over)
* "If the function selected by overload resolution can be determined without instantiating a class template definition, it is unspecified whether that instantiation actually takes place." [[temp.inst]](https://timsong-cpp.github.io/cppwp/temp.inst#7)
* Thus, the compiler is free to instantiate `create_iterable_from_segment(FrameOfReferenceSegment<long>)` even if it will not be called.
* That instantiation, however, will fail because of the static_assert.
* The enable_if version does not suffer from this problem as SFINAE prevents the above method to be added to the resolution candidate set: "If, for a given function template, argument deduction fails or the synthesized function template specialization would be ill-formed, no such function is added to the set of candidate functions for that template."

Long story short: The enable_if should stay, maybe you could add a short comment on what it does (without diving to deep into the details).
